docs/TOFS_pattern.rst

```diff
@@ -0,0 +1,518 @@
+       - 2.ch.pool.ntp.org
+       - 3.ch.pool.ntp.org
+
+Template Override
+^^^^^^^^^^^^^^^^^
+
+If the customization based on pillar data is not enough, we can override the template by creating a new one in ``/srv/saltstack/salt/ntp/files/default/etc/ntp.conf.jinja``
+
+.. code-block:: jinja
+
+   {#- /srv/saltstack/salt/ntp/files/default/etc/ntp.conf.jinja #}
+   {#- Managed by saltstack #}
+   {#- Edit pillars or override this template in saltstack if you need customization #}
+
+   {#- Some bizarre configurations here #}
+   {#- ... #}
+
+   {%- for server in settings.get('servers', default_servers) %}
+   server {{ server }}
+   {%- endfor %}
+
+This way we are locally **overriding the template files** offered by the formula in order to make a more complex adaptation. Of course, this could be applied as well to any of the files, including the state files.
+
+Files Switch
+^^^^^^^^^^^^
+
+To bring some order into the set of template files included in a formula, as we commented, we suggest having a similar structure to a normal final file system under ``files/default``.
+
+We can make different templates coexist for different minions, classified by any `grain <http://docs.saltstack.com/en/latest/topics/targeting/grains.html>`_ value, by simply creating new directories under ``files``. This mechanism is based on **using values of some grains as a switch for the directories under** ``files/``.
+
+If we decide that we want ``os_family`` as switch, then we could provide the formula template variants for both the ``RedHat`` and ``Debian`` families.
+
+.. code-block::
+
+   /srv/saltstack/salt-formulas/ntp-saltstack-formula/ntp/files/
+     default/
+       etc/
+         ntp.conf.jinja
+     RedHat/
+       etc/
+         ntp.conf.jinja
+     Debian/
+       etc/
+         ntp.conf.jinja
+
+To make this work we need a ``conf.sls`` state file that takes a list of possible files as the configuration template.
+
+.. code-block:: sls
+
+   ## /srv/saltstack/salt-formulas/ntp-saltstack-formula/ntp/conf.sls
+   include:
+     - ntp
+
+   {%- from 'ntp/map.jinja' import ntp with context %}
+
+   Configure NTP:
+     file.managed:
+       - name: {{ ntp.config }}
+       - template: jinja
+       - source:
+         - salt://ntp/files/{{ grains.get('os_family', 'default') }}/etc/ntp.conf.jinja
+         - salt://ntp/files/default/etc/ntp.conf.jinja
+       - watch_in:
+         - service: Enable and start NTP service
+       - require:
+         - pkg: Install NTP package
+
+If we want to cover the possibility of a special template for a minion identified by ``pycharm01`` then we could have a specific template in ``/srv/saltstack/salt/ntp/files/pycharm01/etc/ntp.conf.jinja``.
+
+.. code-block:: jinja
+
+   {#- /srv/saltstack/salt/ntp/files/pycharm01/etc/ntp.conf.jinja #}
+   {#- Managed by saltstack #}
+   {#- Edit pillars or override this template in saltstack if you need customization #}
+
+   {#- Some crazy configurations here for pycharm01 #}
```

Comment:
We use the same `docs/TOFS_pattern.rst` in all formulas, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-bf838b5bad3aeeb4d43d3e986622d832R263-R273